README.md
Happit
======

Hackathon Disrupt Europe


grimace-v1.1/CHANGELOG
Version 1.0.1
	- Fixed minor bug in demo.js: wrong path to SaveFile.php
Version 1.1 - 2013-03-20
	- Updated face data
	- New API method loadFromXML allows loading directly from XML objects
	- Facemap can save muscle tensions to clipboard


python/csvtobubble.py
#!/usr/bin/python
from __future__ import print_function
import csv
import sys
import datetime

title = ""
html_file_name = "personal.html"

def getType(string):
    if not string:
        return "number"
    try:
        float(string)
        return "number"
    except ValueError:
        return "string"

def csv2bubble(csvfile, title, delimiter=',', quotechar='"'):
    """
csv to motion
    """

    columns_as_javascript = []
    datarows_as_javascript = []

    csvfile = csv.reader(csvfile, delimiter=delimiter, quotechar=quotechar)
    try:
        columns = csvfile.next()
        first_row = csvfile.next()

        columns_as_javascript.append("data.addColumn('%s', '%s');" % ("string", "ID"))
        columns_as_javascript.append("data.addColumn('%s', '%s');" % ("date", "date"))

        for column_number in range(len(columns))[2:]:
            column_type = getType(first_row[column_number])
            column_name = columns[column_number].replace("_", " ").title()
            columns_as_javascript.append("data.addColumn('%s', '%s');" % (column_type, column_name))

        def row2javascript(row):
            attributes = []
            try:
                datetime.datetime.strptime("2014-06-01", '%Y-%m-%d')
                attributes.append("'%s'" % row[0])  # Id
            except ValueError:
                attributes.append("'1970-01-01'" % row[0])  # Id
            attributes.append("new Date('%s')" % row[1])  # Date

            for column_number in range(len(columns))[2:]:
                # If string surround with '', if number append as it is, if type unknown don't append
                try:
                    if getType(first_row[column_number]) == "string":
                        attributes.append("'%s'" % row[column_number])

                    elif getType(first_row[column_number]) == "number":
                        if row[column_number] == "":
                            attributes.append("NaN")
                        else:
                            attributes.append(str(row[column_number]))
                except IndexError:
                    attributes.append("")

            attributes_joined = ", ".join(attributes)
            datarows_as_javascript.append("\t[%s]" % attributes_joined)

        row2javascript(first_row)
        for row in csvfile:
            row2javascript(row)

        # Insert in HTML

        html = "<!DOCTYPE html>\n<html><head>\n"
        html += "<title>%s</title>" % title
        html += """
        <script type="text/javascript" src="https://www.google.com/jsapi"></script>
        <script type="text/javascript">
        google.load('visualization', '1', {'packages':['motionchart']});
        google.setOnLoadCallback(drawChart);
        function drawChart() {
        var data = new google.visualization.DataTable();
        """
        html += "\n"
        html += "\n".join(columns_as_javascript)
        html += "\n\n"
        html += "data.addRows([\n"+ ",\n".join(datarows_as_javascript)  + "\n]);"
        html += "\n\n"
        html += """
        // This should be a json object INSIDE A STRING.
        var options = {}
        options['state'] = '{}';
        options['width'] = 1000;
        options['height'] = 700;

        var chart = new google.visualization.MotionChart(document.getElementById('chart_div'));


        chart.draw(data, options);
        }
        </script>
        </head>

        <body>"""

        html += "<h1>%s</h1>" % title
        html += """
        <div id="chart_div" style="width: 800px; height: 600px;"></div>
        </body>
        </html>"""

    except StopIteration:
        print("The CSV file should have at least an header and one row.", file=sys.stderr)

    return html

if __name__ == '__main__':
    if not sys.stdin.isatty():
        csvfile = sys.stdin
    elif sys.argv >= 2:
        csvfile = open(sys.argv[1], "r")
    else:
        print("Exmample: cat file.csv | ./csv2mothionchart.py > motionchart.html", file=sys.stderr)
        print("Or: ./csv2mothionchart.py file.csv > motionchart.html", file=sys.stderr)
        sys.exit(1)

    title = "How Others Feel"
    print(csv2motionchart(csvfile, title))


grimace-v1.1/SaveFile.php
<?php

$DIRNAME = 'uploads/';
$PREFIX = '';

$WATCHDOG = 100000;

if (!file_exists($DIRNAME)) {
	mkdir($DIRNAME);
}

// sense mime type of uploaded file for correct extension
$mime = $_SERVER['CONTENT_TYPE'];
switch ($mime) {
	case 'image/png':
		$filetype = 'png';
		break;
		
	case 'image/jpeg':
		$filetype = 'jpg';
		break;
		
	default: 
		echo "Supplied MIME type not allowed for upload.";
		exit(1);
}

// generate filename
do {
	$filename = md5(microtime());
	$uploadUrl = $DIRNAME . $PREFIX . $filename . '.' . $filetype;
}
while(file_exists($uploadUrl));

// handle upload
$postdata = fopen("php://input", "rb"); 
$fp = fopen($uploadUrl, 'wb');

while (!feof($postdata)) {
	fwrite($fp, fread($postdata, 4096));
}

fclose($fp);
fclose($postdata);


// generate absolute URL to uploaded file
$scriptDir = $_SERVER['SCRIPT_NAME'];
$scriptDir = substr($scriptDir,0,strripos($scriptDir,'/'));


printf("http://%s:%s/%s/%s", $_SERVER['SERVER_NAME'], $_SERVER['SERVER_PORT'], $scriptDir, $uploadUrl);
exit(0);




// courtesy of http://at2.php.net/manual/de/function.mkdir.php
/**
 * Makes directory and returns true if exists OR made.
 *
 * @param  $path Path name
 * @return bool
 */
function rmkdir($path, $mode = 0755) {
    $path = rtrim(preg_replace(array("/\\\\/", "/\/{2,}/"), "/", $path), "/");
    $e = explode("/", ltrim($path, "/"));
    if(substr($path, 0, 1) == "/") {
        $e[0] = "/".$e[0];
    }
    $c = count($e);
    $cp = $e[0];
    for($i = 1; $i < $c; $i++) {
        if(!is_dir($cp) && !@mkdir($cp, $mode)) {
            return false;
        }
        $cp .= "/".$e[$i];
    }
    return @mkdir($path, $mode);
}

?>

python/csv2motionchart.py
#!/usr/bin/python
from __future__ import print_function
import csv
import sys
import datetime

title = "Google Motion Chart csv converter Test"
html_file_name = "personal.html"

def getType(string):
    if not string:
        return "number"
    try:
        float(string)
        return "number"
    except ValueError:
        return "string"

def csv2motionchart(csvfile, title, delimiter=',', quotechar='"'):
    """
    This function takes in an csv file and a title and returns a html file with a working Google Motion Chart.
    """


    columns_as_javascript = []
    datarows_as_javascript = []
    
    csvfile = csv.reader(csvfile, delimiter=delimiter, quotechar=quotechar)
    try:
        columns = csvfile.next()
        first_row = csvfile.next()

        columns_as_javascript.append("data.addColumn('%s', '%s');" % ("string", "ID"))
        columns_as_javascript.append("data.addColumn('%s', '%s');" % ("date", "date"))

        for column_number in range(len(columns))[2:]:
            # Google Motion Charts requires ID to be the first field and date to be the second.
            column_type = getType(first_row[column_number])
            column_name = columns[column_number].replace("_", " ").title()
            columns_as_javascript.append("data.addColumn('%s', '%s');" % (column_type, column_name))

        def row2javascript(row):
            attributes = []
            try:
                datetime.datetime.strptime("2014-06-01", '%Y-%m-%d')
                attributes.append("'%s'" % row[0])  # Id
            except ValueError:
                attributes.append("'1970-01-01'" % row[0])  # Id
            attributes.append("new Date('%s')" % row[1])  # Date

            for column_number in range(len(columns))[2:]:
                # If string surround with '', if number append as it is, if type unknown don't append
                try:
                    if getType(first_row[column_number]) == "string":
                        attributes.append("'%s'" % row[column_number])

                    elif getType(first_row[column_number]) == "number":
                        if row[column_number] == "":
                            attributes.append("NaN")
                        else:
                            attributes.append(str(row[column_number]))
                except IndexError:
                    # Google Motion Chart will treat "" as a null value and ether remove bobles where it has no data.
                    # Or interplolate if it as      an value for this field before and after.
                    attributes.append("")

            # data_javascript rows should be on this format:
            # ['Apples',  new Date('2011-04-11'), 1000, 300, 'East'],

            attributes_joined = ", ".join(attributes)
            datarows_as_javascript.append("\t[%s]" % attributes_joined)

        row2javascript(first_row)
        for row in csvfile:
            row2javascript(row)

        # Insert in HTML

        html = "<!DOCTYPE html>\n<html><head>\n"
        html += "<title>%s</title>" % title
        html += """
        <script type="text/javascript" src="https://www.google.com/jsapi"></script>
        <script type="text/javascript">
        google.load('visualization', '1', {'packages':['motionchart']});
        google.setOnLoadCallback(drawChart);
        function drawChart() {
        var data = new google.visualization.DataTable();
        """
        html += "\n"
        html += "\n".join(columns_as_javascript)
        html += "\n\n"
        html += "data.addRows([\n"+ ",\n".join(datarows_as_javascript)  + "\n]);"
        html += "\n\n"
        html += """
        // This should be a json object INSIDE A STRING.
        var options = {}
        options['state'] = '{"iconType":"BUBBLE","dimensions":{"iconDimensions":["dim0"]},"xAxisOption":"10","yZoomedDataMin":-656,"uniColorForNonSelected":false,"sizeOption":"2","showTrails":true,"yLambda":0,"yZoomedIn":false,"xZoomedDataMax":801083,"orderedByY":false,"iconKeySettings":[],"duration":{"multiplier":1,"timeUnit":"D"},"nonSelectedAlpha":0.4,"xZoomedDataMin":5,"colorOption":"3","time":"2014-07-09","yZoomedDataMax":47543,"playDuration":15000,"yAxisOption":"51","xZoomedIn":false,"xLambda":0,"orderedByX":false}';
        options['width'] = 1000;
        options['height'] = 700;

        var chart = new google.visualization.MotionChart(document.getElementById('chart_div'));


        chart.draw(data, options);
        }
        </script>
        </head>

        <body>"""

        html += "<h1>%s</h1>" % title
        html += """
        <div id="chart_div" style="width: 800px; height: 600px;"></div>
        </body>
        </html>"""

    except StopIteration:
        print("The CSV file should have at least an header and one row.", file=sys.stderr)

    return html
            
if __name__ == '__main__':
    if not sys.stdin.isatty():
        csvfile = sys.stdin
    elif sys.argv >= 2:
        csvfile = open(sys.argv[1], "r")
    else:
        print("Exmample: cat file.csv | ./csv2motonchart.py > personal.html", file=sys.stderr)
        print("Or: ./csv2motionchart.py file.csv > motionchart.html", file=sys.stderr)
        sys.exit(1)

    title = "personal"
    print(csv2motionchart(csvfile, title))
        

grimace-v1.1/demos/js/se.php
SUBMITTED <?php
//$file="/var/www/html/grimace-v1.1/demos/js/q.txt";
//$person = "John Smith\n";
//file_put_contents($file, $person, FILE_APPEND | LOCK_EX);
//echo ' :)';
//print_r(error_get_last(),true);
//echo(error_get_last());


/*
echo 'a fear=' . htmlspecialchars($_GET["Xfear"]) . '!';
echo 'QUERY=' . $_GET . '';

echo '  ISSET=' . (isset($_GET['Xjoy']));

echo 'xx';
if (isset($_GET['Xjoy'])){
echo '+++';
}
*/


$myFile = "./recordsfile.html";
//echo '(' . $myFile . ')';

$fh = fopen($myFile, 'a');

//print_r(error_get_last(),true);
/*
echo '[' .$fh . '] file';
if($fh){
	fclose($fh);
echo '(closed.)';
}else{
echo 'FAILED OPENING FILE';

print_r(error_get_last(),true);
}
*/
/* or {echo ("can't open file");};
*/

if($fh){
//$stringData = 'EMOTIONS' . $_GET . "\n";
$stringData = 'EMOTIONS  ' . $_SERVER['QUERY_STRING'] . "\n";
fwrite($fh, $stringData);
//$stringData = "New Stuff 2\n";
//fwrite($fh, $stringData);
fclose($fh);
//echo '[' .$fh . '] file';
//if($fh){
//        fclose($fh);
echo 'SUCCESS.';
}else{
echo 'FAILED OPENING FILE';
//print_r(error_get_last(),true);
}

/*
X_joy='0';
if (isset($_GET["Xjoy"]))){
X_joy=htmlspecialchars($_GET["Xjoy"]) . '';
}

echo 'anyway JOY=' . X_joy;
*/
/*
return;
echo $_GET["Xjoy"]) . '';
echo '. ';
echo $_GET["Xsurprise"]) . '';
echo '. ';
echo $_GET["Xfear"]) . '';
echo '. ';
echo $_GET["Xsadness"]) . '';
echo '. ';
echo $_GET["Xdisgust"]) . '';
echo '. ';
echo $_GET["Xanger"]) . '';
echo '. ';

*/

//phpinfo();
?>


grimace-v1.1/demos/js/demo.js
// portions courtesy of http://livedocs.adobe.com/flash/9.0/ActionScriptLangRefV3/flash/external/ExternalInterface.html
var grimace;
var mysession;

var jsReady = false;
function isReady() {
	return jsReady;
}


function getMovie(movieName) {
  return document[movieName] || window[movieName];
}


function initListeners() {
	try {
		if (!grimace.isReady()) {
			throw new Error('Grimace not ready');
		}
		grimace.addEventListener('loadComplete', 'onLoadComplete');
		grimace.addEventListener('captureComplete', 'onCaptureComplete');
	}
	catch (e) {
		// Retry later if Grimace was not ready yet.
		setTimeout(initListeners, 300);
	}
}



function loadFacedata() {
	var ok;
	try {
		if (!grimace.isReady()) {
			throw new Error('Grimace not ready');
		}
		
		// Paths are relative to location of Grimace.swf
		grimace.loadFacedata([
			'head.xml',
			'features.xml',
			'wrinkles.xml',
			'emotions.xml',
			'overlays.xml'
		], './meta/');
		
		grimace.setCaptureUrl('./SaveFile2.php');
	}
	catch (e) {
		// Retry later if Grimace was not ready yet.
		setTimeout(loadFacedata, 300);
	}
}

function onLoadComplete() {
	document.getElementById('grimace_wheel').style.backgroundImage = 'none';
	grimace.setPosition(0,0);
	grimace.setScaleMode("showAll");

// grimace.setEmotion({anger:Math.random(), surprise:Math.random()}, 0.3);

	grimace.draw();


}

function onCaptureComplete(e) {
	document.getElementById('captures').innerHTML += "<li><img src='"+e.url+"'/></li>";
}



// emotion sliders based on jquery ui
var sliders = {
	resolution: 100,
	
	values: {
		joy:      0,
		surprise: 0,
		fear:     0,
		sadness:  0,
		disgust:  0,
		anger:    0
	},
	
	init: function() {
		var scope = this;
		
		$('#joy').slider({
			max: this.resolution,
			slide: function(e,ui){
				scope.slide('joy');
			}
		});
		$("#surprise").slider({
			max: this.resolution,
			slide: function(e,ui){
				scope.slide('surprise');
			}
		});
		$("#fear").slider({
			max: this.resolution,
			slide: function(e,ui){
				scope.slide('fear');
			}
		});
		$("#sadness").slider({
			max: this.resolution,
			slide: function(e,ui){
				scope.slide('sadness');
			}
		});
		$("#disgust").slider({
			max: this.resolution,
			slide: function(e,ui){
				scope.slide('disgust');
			}
		});
		$("#anger").slider({
			max: this.resolution,
			slide: function(e,ui){
				scope.slide('anger');
			}
		});
	},
	
	notify: true,
	
	slide: function(emotion) {
		if (!this.notify) {
			return;
		}
		
		var value = $('#'+emotion).slider('value');
		value /= this.resolution;

		this.values[emotion] = value;
		
		
		this.update();
		
		// Immediately set new emotion
		grimace.setEmotion(this.values, 0);
	},
	
	update: function() {
		// Deactivate unused sliders when two emotions are already active.
		// Mixtures of more than 3 emotions are possible,
		// but usually unintelligible.
		
		var active = 0;
		for (var i in this.values) {
			if (this.values[i] > 0) {
				active++;
			}
		}
		
		if (active >= 2) {
			for (var i in this.values) {
				if (this.values[i] == 0) {
					$('#'+i).slider('disable');
				}
				else {
					$('#'+i).slider('enable');
				}
			}
		}
		else {
			for (var i in this.values) {
				$('#'+i).slider('enable');
			}
		}
	}
}

function onEmotionSet() {
	grimace.removeEventListener('emotionSet', 'onEmotionSet');
	
	sliders.notify = false;
	
	var emotions = grimace.getEmotion()
	for (var i in sliders.values) {
		if (emotions[i]) {
			sliders.values[i] = emotions[i];
		}
		else {
			sliders.values[i] = 0;
		}
	}
	for (var i in sliders.values) {
		$('#'+i).slider('moveTo', 100 * sliders.values[i]);
	}
	
	sliders.update();
	sliders.notify = true;
}


// Add button functionality
function initButtons() {
	var restart = document.getElementById('restart');
	restart.onclick = function() {
		grimace.restart();
		loadFacedata();
	}
	if (restart.captureEvents) restart.captureEvents(Event.CLICK);

	var draw = document.getElementById('draw');
	draw.onclick = function() {
		grimace.draw();
	}
	if (draw.captureEvents) draw.captureEvents(Event.CLICK);

	var setPos = document.getElementById('setPos');
	setPos.onclick = function() {
		var posX = document.getElementById('posX').value;
		var posY = document.getElementById('posY').value;
		grimace.setPosition(posX, posY);
	}
	if (setPos.captureEvents) setPos.captureEvents(Event.CLICK);

	var getPos = document.getElementById('getPos');
	getPos.onclick = function() {
		var pos = grimace.getPosition();
		var out = 'x: ' + pos.x + '<br/>y: ' + pos.y;
		log(out);
	}
	if (setPos.captureEvents) setPos.captureEvents(Event.CLICK);

	var resize = document.getElementById('resize');
	resize.onclick = function() {
		var width = document.getElementById('width').value;
		var height = document.getElementById('height').value;
		grimace.setMaxBounds(width, height);
	}
	if (resize.captureEvents) resize.captureEvents(Event.CLICK);

	var getBounds = document.getElementById('getBounds');
	getBounds.onclick = function() {
		var bounds = grimace.getMaxBounds();
		var out = 'width: ' + bounds.width + '<br/>height: ' + bounds.height;
		log(out);
	}
	if (getBounds.captureEvents) getBounds.captureEvents(Event.CLICK);

	var set1 = document.getElementById('set1');
	set1.onclick = function() {
		grimace.addEventListener('emotionSet', 'onEmotionSet');
		grimace.setEmotion({anger:1.0, surprise:0.5}, 0.3);
	}
	if (set1.captureEvents) set1.captureEvents(Event.CLICK);

	var set2 = document.getElementById('set2');
	set2.onclick = function() {
		grimace.addEventListener('emotionSet', 'onEmotionSet');
		grimace.setEmotion({joy:1.0, disgust:0.5}, 0.3);
	}
	if (set2.captureEvents) set2.captureEvents(Event.CLICK);

	var reset = document.getElementById('reset');
	reset.onclick = function() {
		grimace.addEventListener('emotionSet', 'onEmotionSet');
		//grimace.resetEmotion(0.3); //NEUTRAL

if(grimace){		
		grimace.setEmotion({anger:Math.round(Math.random()*100)/100/2, surprise:Math.round(Math.random()*100)/100}, 0.3);
}
	//	var r=Math.random();
	//	grimace.setEmotion({r, 1-r}, 0.3);


	}
	if (reset.captureEvents) reset.captureEvents(Event.CLICK);





        //var uname = document.getElementById('uname');
        set2.onclick = function() {
                //grimace.addEventListener('emotionSet', 'onUnameSet');
		var uname = document.getElementById('uname');
                mysession.uname=uname.value;
		//var uname = document.getElementById('uname');
        }
        if (set2.captureEvents) set2.captureEvents(Event.CLICK);














	var getEmotion = document.getElementById('getEmotion');
	getEmotion.onclick = function() {
  	// *************************
		var out = '';
		var emotion = grimace.getEmotion();
		for (var i in emotion) {
			out += i+": "+emotion[i]+"<br/>";
		}
		if (!out) {
			out = "neutral";
		}
		//log(out);

		var outpairs = '';
		for(var i in emotion) {
		   outpairs += ("E_"+i+"="+emotion[i]+"&");
		}
		if(!outpairs)
			outpairs="&";

		 var locxel = document.getElementById("locx");
       		var locyel = document.getElementById("locy");
        	//locxel.value=position.coords.latitude +"";
        	//locyel.value=position.coords.longitude +"";		
		lx=locxel.value;
		ly=locxel.value;

		//mysession.q="2";
		log2(emotion,outpairs,mysession,lx,ly);

		//  grimace.setEmotion({joy:1.0, disgust:0.5}, 0.3);

		//reset.onclick(); //randomize

		// ******************************
	}
	if (getEmotion.captureEvents) getEmotion.captureEvents(Event.CLICK);
	
	var capture = document.getElementById('capture');
	capture.onclick = function() {
		grimace.capture();
	}
	if (capture.captureEvents) capture.captureEvents(Event.CLICK);
	
	var showAll = document.getElementById('showAll');
	showAll.onclick = function() {
		grimace.setScaleMode('showAll');
		document.getElementById('sizeOptions').style.display = 'none';
		grimace.setMaxBounds(0,0);
	}
	if (showAll.captureEvents) showAll.captureEvents(Event.CLICK);
	
	var noScale = document.getElementById('noScale');
	noScale.onclick = function() {
		grimace.setScaleMode('noScale');
		document.getElementById('sizeOptions').style.display = 'block';
	}
	if (noScale.captureEvents) noScale.captureEvents(Event.CLICK);	

//reset.onclick(); //randomize






}

function log2(emotions,ps,mysession,lx,ly){
    var uname = document.getElementById('uname').value;
    var ss;
    //ss="un="+mysession.uname;
    ss="un="+uname+"&";


    // grimace.setEmotion({anger:Math.random(), surprise:Math.random()}, 0.3);


    ls="locx="+lx+"&locy="+ly+"&";

    var xhReq = new XMLHttpRequest();
    xhReq.open("GET", "se.php?"+ss+ps+ls, false); //eubmit emotion

    xhReq.send(null);
    var serverResponse = xhReq.responseText;
//    alert(ps+" "+serverResponse); // Shows "15"
    //
    //document.getElementById('log').innerHTML = msg;

    var stts = document.getElementById("status");
    stts.value = serverResponse;


/*
//WEATHER code
$(document).ready(function(){	
	api_key = 'e85f3f8fd1abe66f'; //http://api.wunderground.com
	weather_getData();
});
*/
}

function weather_getData(){
  $.ajax({
	url: 'http://api.wunderground.com/api/e85f3f8fd1abe66f/forecast/q/37.776289,-122.395234.json',
    dataType: "json",
    success: weather_displayResults
  });
}
function weather_displayResults(data) {
	alert("Temperature: " + current_observation.estimated.temp_c);
}




function log(msg) {
// var str = encodeURIComponent(key) + '=' + encodeURIComponent(parameters[key]);
// xhReq.open("GET", "sumGet.phtml?figure1=5&figure2=10", false);
// xhReq.send(null);
}


jsReady = true;
grimace = getMovie('Grimace');

sliders.init();

initListeners();

initButtons();

// grimace.setEmotion({anger:Math.random(), surprise:Math.random()}, 0.3);

//reset.onclick(); //randomize



//if (document.location.toString().indexOf('file://') != -1) {
// 	alert('Because of Flash security restrictions, this demo must be served by a web server and accessed by http://. You cannot use the demo from a file:// context.');
//}
//else {
	loadFacedata();
//}



python/csv2mothionchart.py
#!/usr/bin/python
from __future__ import print_function
import csv
import sys
import datetime

title = "Google Motion Chart csv converter Test"
html_file_name = "personal.html"

def getType(string):
    if not string:
        return "number"
    try:
        float(string)
        return "number"
    except ValueError:
        return "string"

def csv2motionchart(csvfile, title, delimiter=',', quotechar='"'):
    """
    This function takes in an csv file and a title and returns a html file with a working Google Motion Chart.
    """


    columns_as_javascript = []
    datarows_as_javascript = []
    
    csvfile = csv.reader(csvfile, delimiter=delimiter, quotechar=quotechar)
    try:
        columns = csvfile.next()
        first_row = csvfile.next()

        columns_as_javascript.append("data.addColumn('%s', '%s');" % ("string", "ID"))
        columns_as_javascript.append("data.addColumn('%s', '%s');" % ("date", "date"))

        for column_number in range(len(columns))[2:]:
            # Google Motion Charts requires ID to be the first field and date to be the second.
            column_type = getType(first_row[column_number])
            column_name = columns[column_number].replace("_", " ").title()
            columns_as_javascript.append("data.addColumn('%s', '%s');" % (column_type, column_name))

        def row2javascript(row):
            attributes = []
            try:
                datetime.datetime.strptime("2014-06-01", '%Y-%m-%d')
                attributes.append("'%s'" % row[0])  # Id
            except ValueError:
                attributes.append("'1970-01-01'" % row[0])  # Id
            attributes.append("new Date('%s')" % row[1])  # Date

            for column_number in range(len(columns))[2:]:
                # If string surround with '', if number append as it is, if type unknown don't append
                try:
                    if getType(first_row[column_number]) == "string":
                        attributes.append("'%s'" % row[column_number])

                    elif getType(first_row[column_number]) == "number":
                        if row[column_number] == "":
                            attributes.append("NaN")
                        else:
                            attributes.append(str(row[column_number]))
                except IndexError:
                    # Google Motion Chart will treat "" as a null value and ether remove bobles where it has no data.
                    # Or interplolate if it as      an value for this field before and after.
                    attributes.append("")

            # data_javascript rows should be on this format:
            # ['Apples',  new Date('2011-04-11'), 1000, 300, 'East'],

            attributes_joined = ", ".join(attributes)
            datarows_as_javascript.append("\t[%s]" % attributes_joined)

        row2javascript(first_row)
        for row in csvfile:
            row2javascript(row)

        # Insert in HTML

        html = "<!DOCTYPE html>\n<html><head>\n"
        html += "<title>%s</title>" % title
        html += """
        <script type="text/javascript" src="https://www.google.com/jsapi"></script>
        <script type="text/javascript">
        google.load('visualization', '1', {'packages':['motionchart']});
        google.setOnLoadCallback(drawChart);
        function drawChart() {
        var data = new google.visualization.DataTable();
        """
        html += "\n"
        html += "\n".join(columns_as_javascript)
        html += "\n\n"
        html += "data.addRows([\n"+ ",\n".join(datarows_as_javascript)  + "\n]);"
        html += "\n\n"
        html += """
        // This should be a json object INSIDE A STRING.
        var options = {}
        options['state'] = '{"iconType":"BUBBLE","dimensions":{"iconDimensions":["dim0"]},"xAxisOption":"10","yZoomedDataMin":-656,"uniColorForNonSelected":false,"sizeOption":"2","showTrails":true,"yLambda":0,"yZoomedIn":false,"xZoomedDataMax":801083,"orderedByY":false,"iconKeySettings":[],"duration":{"multiplier":1,"timeUnit":"D"},"nonSelectedAlpha":0.4,"xZoomedDataMin":5,"colorOption":"3","time":"2014-07-09","yZoomedDataMax":47543,"playDuration":15000,"yAxisOption":"51","xZoomedIn":false,"xLambda":0,"orderedByX":false}';
        options['width'] = 1000;
        options['height'] = 700;

        var chart = new google.visualization.MotionChart(document.getElementById('chart_div'));


        chart.draw(data, options);
        }
        </script>
        </head>

        <body>"""

        html += "<h1>%s</h1>" % title
        html += """
        <div id="chart_div" style="width: 800px; height: 600px;"></div>
        </body>
        </html>"""

    except StopIteration:
        print("The CSV file should have at least an header and one row.", file=sys.stderr)

    return html
            
if __name__ == '__main__':
    if not sys.stdin.isatty():
        csvfile = sys.stdin
    elif sys.argv >= 2:
        csvfile = open(sys.argv[1], "r")
    else:
        print("Exmample: cat file.csv | ./csv2mothionchart.py > personal.html", file=sys.stderr)
        print("Or: ./csv2mothionchart.py file.csv > motionchart.html", file=sys.stderr)
        sys.exit(1)

    title = "personal"
    print(csv2motionchart(csvfile, title))
        